source/Mlos.Notebooks/SmartCacheCPP.ipynb

```diff
@@ -0,0 +1,492 @@
+    "        while (!isConfigReady)\n",
+    "        {\n",
+    "            waitForConfigCondVar.wait(lock);\n",
+    "        }\n",
+    "\n",
+    "        config.Update();\n",
+    "        smartCache.Reconfigure();\n",
+    "    }\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "After each iteration, a TelemetryMessage is sent to the MLOS Agent, and the SmartCache blocks until it receives a new configuration to run the next workload.\n",
+    "By default, the agent is not connected to any optimizer, and will not change the original configuration, so the workload will just run uninterrupted."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Starting an Optimizer\n",
+    "We can now also start an Optimizer service for the MLOS Agent to connect to so that we can actually optimize the parameters for this workload.\n",
+    "As the optimizer is running in a separate process, we need to create a new shell on the running docker container using the following command:\n",
+    "\n",
+    "```shell\n",
+    "docker exec -it mlos-build /bin/bash\n",
+    "```\n",
+    "\n",
+    "Within the container, we now install the Python optimizer service:\n",
+    "```shell\n",
+    "pip install -e source/Mlos.Python/\n",
+    "```\n",
+    "\n",
+    "And run it:\n",
+    "```shell\n",
+    "start_optimizer_microservice launch --port 50051\n",
+    "```\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Connecting the Agent to the Optimizer\n",
+    "Now we can start the agent again, this time also pointing it to the optimizer:\n",
+    "```sh\n",
+    "tools/bin/dotnet out/dotnet/source/Mlos.Agent.Server/obj/AnyCPU/Mlos.Agent.Server.dll \\\n",
+    "    --executable out/cmake/Release/source/Examples/SmartCache/SmartCache \\\n",
+    "    --optimizer-uri http://localhost:50051\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This will run the workload again, this time using the optimizer to suggest better configurations. You should see output both in the terminal the agent is running in and in the terminal the OptimizerMicroservice is running in."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Inspecting results\n",
+    "After (or even while) the optimization is running, we can connect to the optimizer via another GRPC channel.\n",
+    "The optimizer is running within the docker container, but when we started docker, we exposed the port 50051 as the same port 50051 on the host machine (on which this notebook is running). So we can now connect to the optimizer within the docker container at `127.0.0.1:50051`.\n",
+    "This assumes this notebook runs in an environment with the `mlos` Python package installed ([see the documentation](https://microsoft.github.io/MLOS/documentation/01-Prerequisites/#python-quickstart))."
```

Comment:
Do we want to propose running the notebook kernel inside the docker instance as well and exposing another port for the user to interact with?
That way they could even execute the build/exec commands from the notebook as well.

Reply:
I went back and forth on that. Then it would be weird to have the instructions for starting the docker container in the notebook. Also, it would require installing jupyter inside the container. But on the other hand it wouldn't require the students to have mlos installed in two python environments. The students should have a local environment with jupyter already set up from the previous assignment though. I think neither is ideal and both are workable. I might do a very quick runthrough video to show how to do it.

Reply:
Ack.
We may revisit for a future tweak to optimize for project newcomers where we can't assume a prior class setup.